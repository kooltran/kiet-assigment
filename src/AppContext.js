import React, { useContext, useCallback, useReducer } from "react";

export const appData = {
  searchResult: [],
  isLoading: false
};

export const AppContext = React.createContext({});

export const AppContextProvider = props => {
  const { children } = props;
  const [data, dispatch] = useReducer(reducer, appData);
  console.log(data.searchResult, "data");
  return (
    <AppContext.Provider value={{ data, setData: dispatch }}>
      {children}
    </AppContext.Provider>
  );
};

const reducer = (initState, action = {}) => {
  switch (action.type) {
    case "SET_BUTTON_ACTION": {
      const { nasa_id, actionType } = action.payload;
      console.log(initState.searchResult, "initialState");
      return {
        ...initState,
        searchResult: initState.searchResult.map(data =>
          data.nasa_id === nasa_id
            ? {
                ...data,
                action: actionType,
                isEdit: actionType === "edit",
                isLike: actionType === "like",
                isRemove: actionType === "remove"
              }
            : data
        )
      };
    }
    case "SET_SEARCH_DATA":
      return {
        ...initState,
        ...action.payload
      };
    case "SET_LOADING":
      return {
        ...initState,
        isLoading: true
      };
    case "SET_EDIT":
      const { nasa_id } = action.payload;
      return {
        ...initState,
        searchResult: initState.searchResult.map(data =>
          data.nasa_id === nasa_id ? { ...data, ...action.payload } : data
        )
      };
    default:
      return {
        ...initState,
        searchResult: action.payload
      };
  }
};

export const useAppContext = () => {
  const { data, setData } = useContext(AppContext);

  const setAction = useCallback(
    (nextData, actionType) => {
      setData({ payload: nextData, type: actionType });
    },
    [setData]
  );

  return [data, setAction];
};
