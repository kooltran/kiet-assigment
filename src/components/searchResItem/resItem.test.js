import React from "react";
import renderer from "react-test-renderer";
import SearchResItem from "./resItem";
import { AppContextProvider } from "../../AppContext";

const baseProps = {
  title: "abc",
  description: "xyz",
  date_created: "2019-03-11T00:00:00Z"
};

it("should render SearchResItem correctly", () => {
  const tree = renderer
    .create(
      <AppContextProvider>
        <SearchResItem item={baseProps} />
      </AppContextProvider>
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
