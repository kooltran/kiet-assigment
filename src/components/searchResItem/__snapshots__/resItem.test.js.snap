// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render SearchResItem correctly 1`] = `
<div
  className="sc-AxjAm eiqBrL"
>
  <div
    className="sc-AxgMl fFPQpl"
  >
    <h3
      className="sc-AxiKw hmgbTf"
    >
      abc
    </h3>
    <p
      className="sc-AxirZ HlyvJ"
    >
      xyz
    </p>
    <p
      className="sc-AxhCb bSmbpm"
    >
      11 Mar 2019
    </p>
  </div>
  <div
    className="sc-AxheI kWoGOa"
  >
    <button
      className="sc-AxhUy dIiefp"
      onClick={[Function]}
      value="like"
    >
      like
    </button>
    <button
      className="sc-AxhUy dIiefp"
      onClick={[Function]}
      value="remove"
    >
      remove
    </button>
    <button
      className="sc-AxhUy dIiefp"
      onClick={[Function]}
      value="edit"
    >
      edit
    </button>
  </div>
</div>
`;
